snforge_std/src/cheatcodes/message_to_l1.cairo

```diff
@@ -0,0 +1,147 @@
+use core::array::ArrayTrait;
+use core::option::OptionTrait;
+use starknet::testing::cheatcode;
+use starknet::{ContractAddress, EthAddress};
+use super::super::_cheatcode::handle_cheatcode;
+
+/// Creates `MessageToL1Spy` instance that spies on all messages sent to L1
+fn spy_messages_to_l1() -> MessageToL1Spy {
+    let mut message_offset = handle_cheatcode(cheatcode::<'spy_messages_to_l1'>(array![].span()));
+    let parsed_message_offset: usize = Serde::<usize>::deserialize(ref message_offset).unwrap();
+
+    MessageToL1Spy { _message_offset: parsed_message_offset }
+}
+
+/// Raw message to l1 format (as seen via the RPC-API), can be used for asserting the sent messages.
+#[derive(Drop, Clone, Serde)]
+struct MessageToL1 {
+    /// An ethereum address where the message is destined to go
+    to_address: EthAddress,
+    /// Actual payload which will be delivered to L1 contract
+    payload: Array<felt252>
+}
+
+/// A message spy structure allowing to get messages emitted only after its creation.
+#[derive(Drop, Serde)]
+struct MessageToL1Spy {
+    _message_offset: usize
+}
+
+/// A wrapper structure on an array of messages to handle filtering smoothly.
+#[derive(Drop, Serde)]
+struct MessagesToL1 {
+    messages: Array<(ContractAddress, MessageToL1)>
+}
+
+trait MessageToL1SpyTrait {
+    /// Gets all messages given [`MessageToL1Spy`] spies for.
+    fn get_messages(ref self: MessageToL1Spy) -> MessagesToL1;
+}
+
+impl MessageToL1SpyTraitImpl of MessageToL1SpyTrait {
+    fn get_messages(ref self: MessageToL1Spy) -> MessagesToL1 {
+        let mut output = handle_cheatcode(
+            cheatcode::<'get_messages_to_l1'>(array![self._message_offset.into()].span())
+        );
+        let messages = Serde::<Array<(ContractAddress, MessageToL1)>>::deserialize(ref output)
+            .unwrap();
+
+        MessagesToL1 { messages }
+    }
+}
+
+trait MessageToL1FilterTrait {
+    /// Filter messages emitted by a sender of a given [`ContractAddress`]
+    fn sent_by(self: @MessagesToL1, contract_address: ContractAddress) -> MessagesToL1;
+    /// Filter messages emitted by a receiver of a given ethereum address (`felt252`)
+    fn sent_to(self: @MessagesToL1, to_address: EthAddress) -> MessagesToL1;
```

Comment:
Shouldn't this be `received_by`?

Reply:
It implies that the message has been received when in fact it hasn't really